great_expectations/datasource/data_connector/file_path_data_connector.py

```diff
@@ -124,7 +124,23 @@ def get_batch_definition_list_from_batch_request(
             A list of BatchDefinition objects that match BatchRequest
 
         """
-        batch_request_base: BatchRequestBase = cast(BatchRequestBase, batch_request)
+        # TODO WILL 202103 some methods still call this method with BatchRequestBase instead of
+        #  _get_batch_definition_list_from_batch_request(). refactor so that the current method
+        #  is only called with BatchRequest
+        if isinstance(batch_request, BatchRequest):
+            if not (
+                batch_request.data_asset_name
+                and isinstance(batch_request.data_asset_name, str)
+            ):
+                raise TypeError(
+                    f"""The type of a data_asset name must be a string (Python "str").  The type given is
+            "{str(type(batch_request.data_asset_name))}", which is illegal.
+                            """
+                )
+            batch_request_base: BatchRequestBase = cast(BatchRequestBase, batch_request)
+        else:
+            batch_request_base: BatchRequestBase = batch_request
+
```

Comment:
There's some additional logic here that I hope to simplify in a subsequent PR.  Currently there are methods in `checkpoints` (and others) that call `get_batch_definition_list_from_batch_request()` using a `BatchRequest` (where `data_asset_name` can't be `None`) and `BatchRequestBase` (where `data_asset_name` can be `None`).   Being more strict about which objects are passed into this method will eliminate the need for this additional logic.

Reply:
@Shinnnyshinshin `BatchRequestBase` is used purely as an attribute bag for internal inter-method communications purposes.  It should not have anything to do with `BatchRequest`, except that the attributes participating in the internal protocol as a subset of the ones in `BatchRequest`.  It should be possible to eliminate `BatchRequestBase` entirely and simply call methods with the arguments they need.  However, that would require renaming some methods (including interface methods) to not have `batch_request` in them.  For example: `_get_batch_definition_list_from_batch_request` (notice the `_` prefix -- it is important; the method without the leading `_` must not be eliminated and/or renamed as it is essential for `BatchRequest` processing.  I would be very much in favor of this refactor.  Having `BatchRequestBase` was a useful intermediate step toward disambiguating `BatchRequestBase` (which can have NULL attributes) and `BatchRequest` (which cannot have the three principal attributes be equal to NULL), but eliminating all together would be even better.  Happy to discuss this in detail!  Thanks!

Reply:
@Shinnnyshinshin On another point, I do not understand why checking for `data_asset_name` being None has to be removed from the up-front processing.  My understanding is that all three principal parameters, `data_source_name`, `data_connector_name`, and `data_asset_name` must NEVER be allowed to be NULL.  So why not check it up front?

Reply:
@Shinnnyshinshin I am assuming that this question being 12 hours ago was already answered in the flexipod meeting -- i.e., `data_asset_name` must never be NULL.

Reply:
yes this was answered :) thank you